pkg/appgw/rewrites.go

```diff
@@ -0,0 +1,180 @@
+	n "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2021-03-01/network"
+	"github.com/Azure/go-autorest/autorest/to"
+	"k8s.io/klog/v2"
+)
+
+// RewriteRuleSets builds the RewriteRuleSets part of the configBuilder
+func (c *appGwConfigBuilder) RewriteRuleSets(cbCtx *ConfigBuilderContext) error {
+
+	if c.appGw.RewriteRuleSets == nil {
+		c.appGw.RewriteRuleSets = &[]n.ApplicationGatewayRewriteRuleSet{}
+	}
+
+	rewriteRuleSets := removeAGICGeneratedRewriteRuleSets(c.appGw.RewriteRuleSets)
+	rewriteRuleSets = append(rewriteRuleSets, c.getAGICRewriteRuleSets(cbCtx)...)
+
+	c.appGw.RewriteRuleSets = &rewriteRuleSets
+	return nil
+}
+
+// removeAGICGeneratedRewriteRuleSets removes the rewrite rule sets that were generated by AGIC
+func removeAGICGeneratedRewriteRuleSets(currentRewriteRuleSets *[]n.ApplicationGatewayRewriteRuleSet) []n.ApplicationGatewayRewriteRuleSet {
+
+	if len(*currentRewriteRuleSets) == 0 {
+		return *currentRewriteRuleSets
+	}
+
+	var appGwRewriteRuleSets []n.ApplicationGatewayRewriteRuleSet
+
+	for _, rrs := range *currentRewriteRuleSets {
+		if rewriteRuleSetName := *(rrs.Name); !(strings.HasPrefix(rewriteRuleSetName, "crd-")) {
+			appGwRewriteRuleSets = append(appGwRewriteRuleSets, rrs)
+		}
+	}
+
+	return appGwRewriteRuleSets
+}
+
+// getAGICRewriteRuleSets returns all the rewrite rule sets that are referenced in atleast one of the ingress manifests
+func (c appGwConfigBuilder) getAGICRewriteRuleSets(cbCtx *ConfigBuilderContext) []n.ApplicationGatewayRewriteRuleSet {
+
+	type rewriteCRDInfo struct {
+		ingressNamespace string
+		CRDName          string
+	}
+
+	uniqueRewriteRuleSetCRD := map[string]rewriteCRDInfo{}
+
+	// insert all referenced rewrite rule sets into a map to avoid duplicates
+	for _, ingress := range cbCtx.IngressList {
+
+		klog.V(9).Infof("Looking for rewrite-rule-set-crd annotation in %s/%s", ingress.Namespace, ingress.Name)
+		rewriteRuleSetCRDName, err := annotations.RewriteRuleSetCRD(ingress)
+
+		// if there is error fetching CRD name or if the value is "", move onto the next ingress
+		if err != nil || rewriteRuleSetCRDName == "" {
+			continue
+		}
+
+		uniqueRewriteRuleSetCRD[ingress.Namespace+"/"+rewriteRuleSetCRDName] = rewriteCRDInfo{
+			ingressNamespace: ingress.Namespace,
+			CRDName:          rewriteRuleSetCRDName,
+		}
+	}
+
+	var appGwRewriteRuleSet []n.ApplicationGatewayRewriteRuleSet
+
+	// get details of all the unique rewrite rule sets referenced in various ingress manifest
+	for _, rewriteRuleSetCRDInfo := range uniqueRewriteRuleSetCRD {
+
+		rewrite, err := c.k8sContext.GetRewrite(rewriteRuleSetCRDInfo.ingressNamespace, rewriteRuleSetCRDInfo.CRDName)
+
+		if err != nil {
+			klog.Errorf("Error occured while fetching rewrite CRD for %s. Please register CRD.", rewriteRuleSetCRDInfo.CRDName)
+			continue
+		}
+
+		appGwRewriteRuleSet = append(appGwRewriteRuleSet, c.makeRewrite(rewriteRuleSetCRDInfo.CRDName, rewrite))
```

Comment:
two rewrite rules CR in different namespace can have the same name. Can you use "namespace-name" as the name of the rewrite rule set ?